src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -319,8 +322,121 @@ private void NotifyChainedHeaderBehaviorsOnReorg()
             this.logger.LogTrace("(-)");
         }
 
-        private async Task<bool> FullyValidateAndReorgAsyncLocked()
+        private async Task<bool> FullyValidateAndReorgLockedAsync(ChainedHeaderBlock chainedHeaderBlock)
         {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            ChainedHeader oldTip = this.Tip;
+            ChainedHeader newTip = chainedHeaderBlock.ChainedHeader;
+
+
+            ChainedHeader fork = oldTip.FindFork(newTip);
+            Queue<ConsensusStoreTransitionState> consensusStoreTransitionQueue = null;
+
+
+            if (fork == newTip)
+            {
+                // New header is behind current tip this is a bug
+                throw new Exception(); // TODO: what do we do.
+            }
+
+            // If the new block is not on the current chain as our current consensus tip
+            // then rewind consensus tip to the common fork (or earlier because rewind might jump a few blocks back)
+            if (fork != oldTip)
+            {
+                ChainedHeader current = oldTip;
+                consensusStoreTransitionQueue = new Queue<ConsensusStoreTransitionState>();
+
+                while (fork.Height < current.Height)
+                {
+                    ConsensusStoreTransitionState transitionState = await this.consensusRules.RewindAsync().ConfigureAwait(false);
+                    current = this.chainedHeaderTree.GetChainedHeaderBlock(transitionState.BlockHash).ChainedHeader;
+                    consensusStoreTransitionQueue.Enqueue(transitionState);
+                }
+
+                fork = current;
+            }
+
+            ChainedHeader[] newChain = newTip.ToChainedHeaderArray(fork);
+
+            ChainedHeader lastValidatedHeader = null;
+            ValidationContext validationContext = null;
+            List<int> badPeers = null;
+
+            foreach (ChainedHeader nextChainedHeader in newChain)
+            {
+                if (nextChainedHeader.BlockValidationState != ValidationState.PartiallyValidated)
+                {
+                    throw new Exception(); // TODO: what do we do.
+                }
+
+                ChainedHeaderBlock nextChainedHeaderBlock = this.chainedHeaderTree.GetChainedHeaderBlock(nextChainedHeader.HashBlock);
+
+                validationContext = new ValidationContext() { Block = nextChainedHeaderBlock.Block };
```

Comment:
The block here can be null, because we can jump back more than we keep in the memory. So we want to load the block from the block store if it was not in the memory and if we have the block store. For light wallet, we will try to make sure that we will never jump back more than the fork, which will ensure the availability of the blocks in the memory on the new forked branch. We cannot make such an ultimate guarantee for bitcoin, but in practice it will never happen. So we would just need the code that recognises that it happened and the up with the consensus tip before the fork.

Reply:
Actually, I have not very happy with that so we're need to change the design a little bit

Reply:
This is no a valid comment anymore I guess, can you delete it if true